bin/prover-client/src/operators/mod.rs

```diff
@@ -61,13 +61,67 @@ pub trait ProvingOp {
     /// - `db`: A reference to the proof database.
     ///
     /// # Returns
-    /// A tuple containing the primary `ProofContext` and a vector of dependent `ProofContext`s.
+    /// A vector of [`ProofKey`] corresponding to a given proving operation.
     async fn create_task(
         &self,
         params: Self::Params,
         task_tracker: Arc<Mutex<TaskTracker>>,
         db: &ProofDb,
-    ) -> Result<Vec<ProofKey>, ProvingTaskError>;
+    ) -> Result<Vec<ProofKey>, ProvingTaskError> {
+        let proof_ctx = self.construct_proof_ctx(&params)?;
+
+        // Try to fetch the existing prover tasks for dependencies.
+        let proof_deps = db
+            .get_proof_deps(proof_ctx)
+            .map_err(ProvingTaskError::DatabaseError)?;
+
+        let cl_stf_deps = match proof_deps {
+            // Reuse the existing dependency tasks fetched from DB.
+            Some(v) => v,
+            // Create new dependency tasks.
+            None => {
+                let deps_keys = self
+                    .create_deps_tasks(params, db, task_tracker.clone())
+                    .await?;
+                let deps: Vec<_> = deps_keys.iter().map(|v| v.context().to_owned()).collect();
+
+                if !deps.is_empty() {
+                    db.put_proof_deps(proof_ctx, deps.clone())
+                        .map_err(ProvingTaskError::DatabaseError)?;
+                }
+                deps
+            }
+        };
+
+        let mut task_tracker = task_tracker.lock().await;
+        task_tracker.create_tasks(proof_ctx, cl_stf_deps, db)
+    }
+
+    /// Construct [`ProofContext`] from the proving operation parameters.
+    fn construct_proof_ctx(&self, params: &Self::Params) -> Result<ProofContext, ProvingTaskError>;
+
+    /// Creates a set of dependency tasks.
+    ///
+    /// # Important
+    /// The default impl defines no dependencies, so certain [`ProvingOp`] with dependencies
+    /// should "override" it.  
+    ///
+    /// # Arguments
+    /// - `params`: The parameters specific to the operation.
+    /// - `task_tracker`: A shared task tracker for managing task dependencies.
+    /// - `db`: A reference to the proof database.
+    ///
+    /// # Returns
+    /// A [`Vec`] containing the [`ProofKey`] for the dependent proving operations.
```

Comment:
```suggestion
    /// # Important
    ///
    /// The default impl defines no dependencies, so certain [`ProvingOp`] with dependencies
    /// should "override" it.  
    ///
    /// # Arguments
    ///
    /// - `params`: The parameters specific to the operation.
    /// - `task_tracker`: A shared task tracker for managing task dependencies.
    /// - `db`: A reference to the proof database.
    ///
    /// # Returns
    ///
    /// A [`Vec`] containing the [`ProofKey`] for the dependent proving operations.
```

Empty line after headers.

Reply:
done